### PR TITLE
refactor(trie): Separate database integration its own crate

### LINKED ISSUE #8514: Extract reth-trie-types from reth-trie
### Describe the feature

reth-trie mixes db implementations with type definitions, which is problematic if we want to use these types in evm for different targets (risc, maybe even no_std).

hence we should extract the core types to a standalone crate, primarily the TrieUpdates type

currently this type has functions that make use of db:

https://github.com/paradigmxyz/reth/blob/f43ccc760b150854654e7364892f6ec753c17eeb/crates/trie/trie/src/updates.rs#L132-L132

or 

https://github.com/paradigmxyz/reth/blob/f43ccc760b150854654e7364892f6ec753c17eeb/crates/trie/trie/src/updates.rs#L114-L119

The triewalker functions could be inverted and implemented on triewalker instead,
for the flush function we could use a tmp wrapper type instead

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by RomanHodulak**
### About
Closes #8514

- replace `TrieWalker` dependency on database with a trait
- use custom associated error type instead of `DatabaseError` everywhere
- remove dependencies on `reth-db` and `reth-db-api` from `reth-trie`

### Motivation
The motivation for this change is adapted from the linked issue description. Essentially, the `reth-trie` defines the Trie domain. However, in its current form it also depends on `reth_db` database. This makes the `reth-trie` crate responsible for two things at once. On one hand, it defines the Trie domain, on the other, it also depends on a specific implementation of some of its traits. This makes reth-trie inseperable from reth-db and contributes to model being anemic. The practical implication of this is the inability to compile into different targets such as RISC-V.

### Goals
The goal of this change is to separate `reth_db` from `reth_trie`.

### Solution
The solution is to introduce `reth-trie-db` crate that implements `reth_trie` using `reth_db`. You can think of `reth-trie-db` as an anti-corruption layer.

 * `reth-trie`  - contains types and behavior of the Trie domain, but does not depend on a database, the concept of database does not really exist there.
 * `reth-trie-db` - is a projection of the `reth-trie` into the database, but does not define any concepts of the Trie domain. It implements the traits defined by the Trie domain using `reth_db`. There might be other implementations than the `reth_db`. that's why the `reth-trie-db` shouldn't contain domain behavior or types, otherwise that would lead to duplication of knowledge.

### Additional notes

The unit tests in `reth-trie` are also dependent on database. To keep this PR at a reasonable size, the solution here is to port these tests over to the new `reth-trie-db`. Ideally, the unit tests should not be dependent on `reth-db`.

### CODE REVIEW

File: crates/trie/db/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+pub mod cursors;
+pub mod prefix_set;
+pub mod proof;
+pub mod state;
+pub mod trie;
+pub mod walker;
+
+/// New-type that wraps an immutable reference to a transaction.
+#[derive(Debug)]
+pub struct TxRefWrapper<'a, T>(&'a T);
```
Comment (mattsse) [inline crates/trie/db/src/lib.rs:17]: hmm, this required now because we have traits separate?

not a big deal imo

**Review by mattsse — COMMENTED** [review]
tysm for this.

@rkrasiuk I like this a lot.

this makes the trie types easier to use because it separates the the db stuff entirely.

**Comment by rkrasiuk** [comment]
@RomanHodulak appreciate the PR. the motivation is solid and we indeed need this, however, could you please try to split this into smaller PRs. all of this code is in the critical path and having PR as big as is rn would require a couple of hours to review 🙏

**Comment by RomanHodulak** [comment]
Hey @rkrasiuk and @mattsse, it's good to hear from you again.

Thanks for looking over my PR.

I appreciate the feedback and understand the concerns.

Sure I can come up with a way to separate this into smaller PRs.

I'm thinking at least these could be introduced separately and iteratively:

1. Replacing `DatabaseError` with associated error type.
2. Introducing new `reth-trie-db` crate.
3. Adding writable trie cursors.
4. Replacing database types with traits.
5. Moving trie implementations to the db crate.
6. Moving factory methods (such as Proof and StateRoot) based on database transaction to the db crate.
7. Finally, removing `reth-db` and `reth-db-api` dependencies from `reth-trie`.

**Comment by rkrasiuk** [comment]
@RomanHodulak thanks for putting this together. the list looks reasonable albeit the error type replacement. keen to review and get the other PRs merged as they come

**Comment by mattsse** [comment]
@RomanHodulak I think we should start with adding the new reth-trie-db crate (can be empty at first), and then incrementally populate it and make the necessary changes?

**Comment by rkrasiuk** [comment]
@RomanHodulak let's re-align here on the final steps to consider this change complete. once https://github.com/paradigmxyz/reth/pull/9970 and https://github.com/paradigmxyz/reth/pull/9985 are in, the only things that are left are to simply move database implementations of `TrieCursorFactory` and `HashedStateFactory` to `reth-trie-db` crate. anything I'm missing?

**Comment by RomanHodulak** [comment]
@rkrasiuk That's right.

**Comment by rkrasiuk** [comment]
@RomanHodulak great! what are the chances we could finish this by eod?

**Comment by RomanHodulak** [comment]
@rkrasiuk Very high

**Comment by rkrasiuk** [comment]
@RomanHodulak marvelous, looking forward!

**Pull request closed without merging**